newton/examples/ik/example_ik_custom.py

```diff
@@ -0,0 +1,324 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 The Newton Developers
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+###########################################################################
+# Example IK Custom (custom collision objective + sphere gizmo)
+#
+# Inverse kinematics on a Franka arm, with collision
+# avoidance against an interactive sphere obstacle.
+#
+# - Adds a custom CollisionSphereAvoidObjective (softplus penalty) for the EE
+# - Adds gizmos for the end-effector target and the obstacle sphere
+# - Re-solves IK every frame from the latest gizmo transforms
+#
+# Command: python -m newton.examples ik_custom
+###########################################################################
+
+import warp as wp
+
+import newton
+import newton.examples
+import newton.ik as ik
+import newton.utils
+
+
+# -------------------------------------------------------------------------
+# Custom collision residuals
+# -------------------------------------------------------------------------
+@wp.kernel
+def _collision_residuals(
+    body_q: wp.array2d(dtype=wp.transform),  # (batch_rows, n_bodies)
+    obstacle_centers: wp.array1d(dtype=wp.vec3),  # (n_problems,)
+    obstacle_radii: wp.array1d(dtype=wp.float32),  # (n_problems,)
+    link_index: int,
+    link_offset: wp.vec3,
+    link_radius: float,
+    start_idx: int,  # start row in the global residual vector
+    weight: float,
+    problem_idx: wp.array1d(dtype=wp.int32),  # (batch_rows,)
+    # outputs
+    residuals: wp.array2d(dtype=wp.float32),  # (batch_rows, total_residuals)
+):
+    row_idx = wp.tid()
+    base = problem_idx[row_idx]
+
+    # EE sphere centre in world space
+    tf = body_q[row_idx, link_index]
+    ee_pos = wp.transform_point(tf, link_offset)
+
+    # Obstacle sphere (in world space)
+    c = obstacle_centers[base]
+    r_obs = obstacle_radii[base]
+
+    # Softplus of penetration depth to keep it smooth
+    dist = wp.length(ee_pos - c)
+    delta = (link_radius + r_obs) - dist
+    margin = 0.05
+    pen = wp.log(1.0 + wp.exp(delta / margin)) * margin
+
+    residuals[row_idx, start_idx] = weight * pen
+
+
+@wp.kernel
+def _update_center_target(
+    problem_idx: int,
+    new_center: wp.vec3,
+    # outputs
+    centers: wp.array1d(dtype=wp.vec3),  # (n_problems,)
+):
+    centers[problem_idx] = new_center
+
+
+class CollisionSphereAvoidObjective(ik.IKObjective):
```

Comment:
I think it would be good for code linters if we always added type definitions for all method arguments in those user-facing API classes, e.g. `ik.IKObjective`. That way it becomes also easier for the user to build on top of `ik.IKObjective` to know what inputs are expected.